### PR TITLE
ci: enforce LibAsset routing for ERC20 transfers

### PR DESCRIPTION
**Pull request by 0xDEnYO**
# Which Linear task belongs to this PR?

[EXSC-300](https://linear.app/lifi-linear/issue/EXSC-300/add-solhint-rule-or-ci-grep-enforcing-libasset-routing-for-erc20) — Add solhint rule (or CI grep) enforcing LibAsset routing for ERC20 transfers

Audit follow-up to PR #1715 (Sujith, informational finding #4).

# Why did I implement it this way?

The ticket offered two options: a custom solhint rule or a CI grep step. I went with the **CI grep step** (a new workflow, `enforceLibAssetRouting.yml`):

- A custom solhint rule would require shipping a local JS solhint plugin package (`solhint-plugin-*`), which conflicts with the repo's TypeScript/Bash-only scripting rule and adds dependency surface for a single rule.
- The grep backstop is explicitly sanctioned by the ticket as "adequate", is trivially auditable, and follows the established house pattern of inline-bash checker workflows (`jsonChecker.yml`, `protectSecurityRelevantCode.yml`).

Behavior:

- Fails CI if any `.sol` file in `src/` outside the allowlist contains `.safeTransfer(` or `.safeTransferFrom(`.
- `safeTransferETH(` is not matched and remains allowed everywhere (the regex requires `(` directly after `safeTransfer`/`safeTransferFrom`).
- The four packed facets (`HopFacetPacked`, `AcrossFacetPacked`, `CBridgeFacetPacked`, `AcrossFacetPackedV3`) are explicitly allowlisted with a comment, per the AC.
- Beyond the four facets named in the AC, six more pre-existing offenders exist on `main` (`TokenWrapper`, `LiFiDEXAggregator`, `ReceiverStargateV2`, `ReceiverAcrossV3`, `ReceiverChainflip`, `WithdrawablePeriphery` — consistent with the audit finding "Out-of-scope facets and LidoWrapper bypass LibAsset"). They are grandfathered in a separately-commented allowlist section so the check is green on `main` from day one while still rejecting any NEW file or any new call site in a currently-clean file.

Verification evidence (run locally against this branch):

- `bash -n` on the extracted check script: pass
- Clean tree (current `src/`): exit 0
- Injected `src/EvilFacet.sol` with `IERC20(t).safeTransfer(...)`: exit 1, offending file + line printed
- Injected `src/EthOnlyFacet.sol` with only `safeTransferETH` calls (both method-style and library-style): exit 0
- `bunx prettier --check` on the workflow file: pass

## DRAFT status / pr-ready note

`/pr-ready` (local CodeRabbit) could not be run to a clean state: the CLI returned "Review limit reached" (org usage credits exhausted), twice. Per the pre-PR workflow this PR is therefore opened as a **draft**; cloud CodeRabbit will review on this PR as the safety net. A `self-review-pass` sweep (mechanical/semantic/executable) was completed with no findings.

# Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] This pull request is as small as possible and only tackles one problem
- [ ] I have run `/pr-ready` (local CodeRabbit) on this branch and resolved (or explicitly documented) all findings — see `.agents/commands/pr-ready.md` (attempted; rate-limited — documented above)
- [x] I have added tests that cover the functionality / test the bug (manual positive/negative/exemption runs of the check script, evidence above; no Solidity changes)
- [ ] For new facets: I have checked all points from this list: https://www.notion.so/lifi/New-Facet-Contract-Checklist-157f0ff14ac78095a2b8f999d655622e (N/A — no facets touched)
- [x] I have updated any required documentation (workflow header documents purpose, triggers, behavior, limitations)

# Checklist for reviewer (DO NOT DEPLOY and contracts BEFORE CHECKING THIS!!!)

- [ ] I have checked that any arbitrary calls to external contracts are validated and or restricted
- [ ] I have checked that any privileged calls (i.e. storage modifications) are validated and or restricted
- [ ] I have ensured that any new contracts have had AT A MINIMUM 1 preliminary audit conducted on <date> by <company/auditor>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .github/workflows/enforceLibAssetRouting.yml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        uses: actions/checkout@34e114876b0b11c390a56381ad16ebd13914f8d5 # v4
+
+      # Step 2: Scan src/ for direct ERC20 safeTransfer / safeTransferFrom calls
+      - name: Check for direct ERC20 transfers bypassing LibAsset
+        run: |
+          set -euo pipefail
+
+          ##### Files allowed to call .safeTransfer( / .safeTransferFrom( directly.
+          ##### Everything else in src/ must route ERC20 transfers through LibAsset.
+          ALLOWLIST=(
```
Comment (mirooon) [inline .github/workflows/enforceLibAssetRouting.yml:40]: **The hardcoded allowlist is in the wrong place — this is a smart-contract decision, not a CI-infra one**

Because this list lives in the workflow under `.github/`, any edit to it triggers `protectSecurityRelevantCode.yml`, which requires an **InformationSecurityManager** approval. That gate is there to protect our CI integrity (so checks can't be quietly disabled) — it is not the team that reviews contract logic.

But "may this contract bypass LibAsset?" is a **smart-contract security** decision. We review those in normal SC PR review, not through the InfoSec-infra gate. So with the list here, every time someone grandfathers or removes a contract, they pull in the wrong approvers for a decision that should sit with SC reviewers. The list is also disconnected from the contracts it describes, so it goes stale on a rename or deprecation.

I'd suggest moving the list into `config/global.json`, next to the lists we already keep there (`coreFacets`, `corePeriphery`, `blacklistedFunctionSelectors`). For example:

```json
"libAssetRoutingExempt": [
  "AcrossFacetPacked",
  "AcrossFacetPackedV3",
  "CBridgeFacetPacked",
  "HopFacetPacked",
  "WithdrawablePeriphery",
  "LiFiDEXAggregator",
  "ReceiverAcrossV3",
  "ReceiverChainflip",
  "ReceiverStargateV2",
  "TokenWrapper"
]
```

The workflow then reads it with `jq` instead of carrying the array inline. `config/` is not under `.github/`, so edits go through normal SC review (the right people) — while the workflow *logic* stays protected by the InfoSec gate, which is what that gate is actually for. What do you think?

File: .github/workflows/enforceLibAssetRouting.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            "src/Helpers/WithdrawablePeriphery.sol"
+            "src/Periphery/LiFiDEXAggregator.sol"
+            "src/Periphery/ReceiverAcrossV3.sol"
+            "src/Periphery/ReceiverChainflip.sol"
+            "src/Periphery/ReceiverStargateV2.sol"
+            "src/Periphery/TokenWrapper.sol"
+          )
+
+          ##### Matches .safeTransfer( and .safeTransferFrom( but NOT .safeTransferETH(
+          PATTERN='\.safeTransfer(From)?\('
```
Comment (mirooon) [inline .github/workflows/enforceLibAssetRouting.yml]: **Coverage gap: raw `.transfer` / `.transferFrom` are not caught**

The regex only matches `.safeTransfer(` and `.safeTransferFrom(`. It does **not** match raw `.transfer(` / `.transferFrom(`. I checked `src/` and `LidoWrapper.sol` uses the raw form:

```
src/Periphery/LidoWrapper.sol:71   IERC20(address(ST_ETH)).transferFrom(...)
src/Periphery/LidoWrapper.sol:85   IERC20(WST_ETH_ADDRESS).transfer(msg.sender, wrappedAmount);
```

This stood out to me because audit finding #4 is literally titled *"Out-of-scope facets **and LidoWrapper** bypass LibAsset"*. So the one contract the auditor named is not caught by the check **and** is not in the allowlist — it just passes silently.

The bigger risk is the bug class itself: the original Tron USDT problem was a raw `transfer()` returning bad data. With the current regex, a new facet that calls `IERC20(token).transfer(...)` directly would pass CI and could bring that exact bug back.

Two ways to handle it:
1. Widen the pattern to also catch raw transfers, e.g. `\.(safe)?[Tt]ransfer(From)?\(`, and add `LidoWrapper.sol` to the allowlist with a short comment.
2. Keep the narrow scope, but then document in the header that raw `.transfer`/`.transferFrom` is intentionally out of scope and LidoWrapper is a known gap.

I'd lean toward option 1, since it closes the exact bug class this check exists to prevent. What do you think?

File: .github/workflows/enforceLibAssetRouting.yml (hunk 1 above)
Comment (0xDEnYO) [reply .github/workflows/enforceLibAssetRouting.yml:40]: Thanks for your comment. I see your point but I have a different opinion. 
Using a bypass should be avoided at any cost. The exemptions are there only for legacy so we dont need to change a number of already-deployed contracts. 

It should not be easy for a developer to not use LibAsset. If we just allow setting a flag in global.json, we make it rather easy. Having it where it is right now requires the additional security approval which is an additional check.

 I would suggest to leave it where it is, since we wont plan to use this regularly as a feature, it is a legacy exemption only.

File: .github/workflows/enforceLibAssetRouting.yml (hunk 2 above)
Comment (0xDEnYO) [reply .github/workflows/enforceLibAssetRouting.yml]: Went with option 1 — implemented in 4ea9719d:

- Pattern widened to `\.(safeT|t)ransfer(From)?\(` so raw `.transfer(`/`.transferFrom(` are now caught too (still excludes `.safeTransferETH(`).
- `LidoWrapper.sol` added to the allowlist as grandfathered (it's the contract named in audit finding #4), with a comment marking it as not endorsed.
- The widened pattern also flagged `Executor.sol:168` — that one is a false positive: it calls `transferFrom` on our own `ERC20Proxy`, which itself routes through `LibAsset.transferFromERC20`. Allowlisted with a comment explaining that.

Verified locally: the check passes on the current tree, fails on a synthetic raw-`transfer` violation, and all four call forms match while `safeTransferETH(` doesn't.

**Comment by 0xDEnYO** [comment]
@coderabbitai The `persist-credentials: false` nitpick is addressed in 88ad8319.

**Review by mirooon — APPROVED** [review]

**Review by maxklenk — APPROVED** [review]
